security/pkg/k8s/controller/monitoring.go

```diff
@@ -15,78 +15,72 @@
+		"The number of certificates deleted due to service account deletion.",
+	)
 
-	csrErrorCounts = prometheus.NewCounterVec(prometheus.CounterOpts{
-		Namespace: "citadel",
-		Subsystem: "secret_controller",
-		Name:      "csr_err_count",
-		Help:      "The number of errors occurred when creating the CSR.",
-	}, []string{})
+	secretDeletionCounts = monitoring.NewSum(
+		"secret_deleted_cert_count",
+		"The number of certificates recreated due to secret deletion (service account still exists).",
+	)
 
-	certSignErrorCounts = prometheus.NewCounterVec(prometheus.CounterOpts{
-		Namespace: "citadel",
-		Subsystem: "secret_controller",
-		Name:      "csr_sign_err_count",
-		Help:      "The number of errors occurred when signing the CSR.",
-	}, []string{errorlabel})
+	csrErrorCounts = monitoring.NewSum(
+		"csr_err_count",
+		"The number of errors occurred when creating the CSR.",
+	)
+
+	certSignErrorCounts = monitoring.NewSum(
+		"csr_sign_err_count",
+		"The number of errors occurred when signing the CSR.",
+	)
 )
 
 func init() {
-	prometheus.MustRegister(serviceAccountCreationCounts)
-	prometheus.MustRegister(serviceAccountDeletionCounts)
-	prometheus.MustRegister(secretDeletionCounts)
-	prometheus.MustRegister(csrErrorCounts)
-	prometheus.MustRegister(certSignErrorCounts)
+	monitoring.MustRegister(
+		serviceAccountCreationCounts,
+		serviceAccountDeletionCounts,
+		secretDeletionCounts,
+		csrErrorCounts,
+		certSignErrorCounts,
+	)
 }
 
 // monitoringMetrics are counters for secret controller operations.
 type monitoringMetrics struct {
-	ServiceAccountCreation prometheus.Counter
-	ServiceAccountDeletion prometheus.Counter
-	SecretDeletion         prometheus.Counter
-	CSRError               prometheus.Counter
-	certSignErrors         *prometheus.CounterVec
+	ServiceAccountCreation monitoring.Metric
+	ServiceAccountDeletion monitoring.Metric
+	SecretDeletion         monitoring.Metric
+	CSRError               monitoring.Metric
+	certSignErrors         monitoring.Metric
 }
 
 // newMonitoringMetrics creates a new monitoringMetrics.
 func newMonitoringMetrics() monitoringMetrics {
 	return monitoringMetrics{
-		ServiceAccountCreation: serviceAccountCreationCounts.With(prometheus.Labels{}),
-		ServiceAccountDeletion: serviceAccountDeletionCounts.With(prometheus.Labels{}),
-		SecretDeletion:         secretDeletionCounts.With(prometheus.Labels{}),
-		CSRError:               csrErrorCounts.With(prometheus.Labels{}),
+		ServiceAccountCreation: serviceAccountCreationCounts,
+		ServiceAccountDeletion: serviceAccountDeletionCounts,
+		SecretDeletion:         secretDeletionCounts,
+		CSRError:               csrErrorCounts,
 		certSignErrors:         certSignErrorCounts,
 	}
 }
 
-func (m *monitoringMetrics) GetCertSignError(err string) prometheus.Counter {
-	return m.certSignErrors.With(prometheus.Labels{errorlabel: err})
+func (m *monitoringMetrics) GetCertSignError(err string) monitoring.Metric {
+	return m.certSignErrors.With(errorTag.Value(err))
```

Comment:
I don't remember if this will work without adding the tag key during the creation first. Let me double check.

Reply:
See https://github.com/istio/istio/pull/15223#discussion_r305134021